File: typescript/src/live/live-from-file-without-sdk.ts
import { type LiveV2InitRequest, type LiveV2InitResponse } from '@gladiaio/sdk';
import WebSocket from 'ws';
import {
  getAudioFileFormat,
  initFileRecorder,
  printMessage,
  readGladiaApiKey,
} from './helpers.js';

const gladiaApiKey = readGladiaApiKey();
const gladiaApiUrl = 'https://api.gladia.io';
const region = 'eu-west'; // us-west

const filepath = '../data/anna-and-sasha-16000.wav';
const config: LiveV2InitRequest = {
  language_config: {
    languages: ['es', 'ru', 'en', 'fr'],
    code_switching: true,
  },
  messages_config: {
    receive_partial_transcripts: false, // Set to true to receive partial/intermediate transcript
    receive_final_transcripts: true,
  },
};

async function initLiveSession(): Promise<LiveV2InitResponse> {
  const response = await fetch(`${gladiaApiUrl}/v2/live?region=${region}`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'X-GLADIA-KEY': gladiaApiKey,
    },
    body: JSON.stringify({ ...getAudioFileFormat(filepath), ...config }),
  });
  if (!response.ok) {
    console.error(
      `${response.status}: ${(await response.text()) || response.statusText}`,
    );
    process.exit(response.status);
  }

  return (await response.json()) as LiveV2InitResponse;
}

function initWebSocket(
  { url }: LiveV2InitResponse,
  onOpen: () => void,
  onClose: (event: { code: number; reason: string }) => void,
): WebSocket {
  const socket = new WebSocket(url);

  socket.addEventListener('open', function () {
    onOpen();
  });

  socket.addEventListener('error', function (error) {
    console.error(error);
  });

  socket.addEventListener('close', async ({ code, reason }) => {
    onClose({ code, reason });
  });

  socket.addEventListener('message', function (event) {
    // All the messages we are sending are in JSON format
    const message = JSON.parse(event.data.toString());
    printMessage(message);
  });

  return socket;
}

async function start() {
  const initiateResponse = await initLiveSession();

  const socket: WebSocket = new WebSocket(initiateResponse.url);

  const recorder = initFileRecorder(
    // Send every chunk from recorder to the socket
    (chunk) => socket.send(chunk),
    // When the recording is stopped, we send a message to tell the server
    // we are done sending audio and it can start the post-processing
    () => socket.send(JSON.stringify({ type: 'stop_recording' })),
    filepath,
  );

  socket.addEventListener('open', function () {
    console.log();
    console.log(
      `################ Begin session ${initiateResponse.id} ################`,
    );
    console.log();

    recorder.start();
  });

  socket.addEventListener('message', function (event) {
    // All the messages we are sending are in JSON format
    const message = JSON.parse(event.data.toString());
    printMessage(message);
  });

  socket.addEventListener('error', function (error) {
    console.error(error);
  });

  socket.addEventListener('close', async ({ code, reason }) => {
    console.log();
    console.log(
      `################  End session ${initiateResponse.id}  ################`,
    );
    console.log();
  });
}

start();

File: python/src/streaming/live_from_file_without_sdk.py
import asyncio
import json
from typing import Any

import httpx
from websockets.asyncio.client import ClientConnection, connect
from websockets.exceptions import ConnectionClosedOK

from env import get_gladia_api_key, get_gladia_region, get_gladia_api_url
from file import get_data_file_path

from .helper import print_message

audio_file = "anna-and-sasha-16000.wav"
sample_rate = 16_000
bit_depth = 16
channels = 1

STREAMING_CONFIGURATION = {
    "encoding": "wav/pcm",
    "sample_rate": sample_rate,
    "bit_depth": bit_depth,
    "channels": channels,
    "language_config": {
        "languages": ["es", "ru", "en", "fr"],
        "code_switching": True,
    },
    "messages_config": {
        "receive_partial_transcripts": False,  # Set to True to receive partial/intermediate transcript
        "receive_final_transcripts": True,
    },
}


async def run():
    response = init_live_session(STREAMING_CONFIGURATION)
    async with connect(response["url"]) as websocket:
        tasks: list[asyncio.Task[Any]] = []
        try:
            print(f"\n################ Begin session {response['id']} ################\n")
            tasks.append(asyncio.create_task(send_audio(websocket)))
            tasks.append(asyncio.create_task(receive_messages_from_socket(websocket)))

            _ = await asyncio.wait(tasks)
        except asyncio.exceptions.CancelledError:
            for task in tasks:
                _ = task.cancel()
    print(
        f"\n################ End session {response['id']} ################\n"
    )


async def receive_messages_from_socket(socket: ClientConnection) -> None:
    async for message in socket:
        print_message(str(message))


def init_live_session(config: dict[str, Any]) -> dict[str, Any]:
    gladia_key = get_gladia_api_key()
    return httpx.post(
        f"{get_gladia_api_url()}/v2/live",
        params={"region": get_gladia_region()},
        headers={"X-Gladia-Key": gladia_key},
        json=config,
        timeout=3,
    ).json()


async def send_audio(socket: ClientConnection) -> None:
    file_path = get_data_file_path(audio_file)
    with open(file_path, "rb") as f:
        file_content = f.read()
    file_content = file_content[44:]  # Skip the WAV header

    chunk_duration = 0.1
    chunk_size = int(
        sample_rate * (bit_depth / 8) * channels * chunk_duration
    )

    # Send the audio file in chunks
    offset = 0
    while offset < len(file_content):
        try:
            await socket.send(file_content[offset : offset + chunk_size])
            offset += chunk_size
            await asyncio.sleep(chunk_duration)
        except ConnectionClosedOK:
            return
    print(">>>>> Sent all audio data")
    await socket.send(json.dumps({"type": "stop_recording"}))


def main():
    asyncio.run(run())


if __name__ == "__main__":
    main()
